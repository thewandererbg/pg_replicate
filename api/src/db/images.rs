use sqlx::{PgPool, Postgres, Transaction};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ImagesDbError {
    #[error("Error while interacting with PostgreSQL for images: {0}")]
    Database(#[from] sqlx::Error),
}

pub struct Image {
    pub id: i64,
    pub name: String,
    pub is_default: bool,
}

pub async fn create_image(
    pool: &PgPool,
    name: &str,
    is_default: bool,
) -> Result<i64, ImagesDbError> {
    let mut txn = pool.begin().await?;
    let res = create_image_txn(&mut txn, name, is_default).await;
    txn.commit().await?;
    res
}

pub async fn create_image_txn(
    txn: &mut Transaction<'_, Postgres>,
    name: &str,
    is_default: bool,
) -> Result<i64, ImagesDbError> {
    let record = sqlx::query!(
        r#"
        insert into app.images (name, is_default)
        values ($1, $2)
        returning id
        "#,
        name,
        is_default
    )
    .fetch_one(&mut **txn)
    .await?;

    Ok(record.id)
}

pub async fn read_default_image(pool: &PgPool) -> Result<Option<Image>, ImagesDbError> {
    let record = sqlx::query!(
        r#"
        select id, name, is_default
        from app.images
        where is_default = true
        "#,
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| Image {
        id: r.id,
        name: r.name,
        is_default: r.is_default,
    }))
}

pub async fn read_image(pool: &PgPool, image_id: i64) -> Result<Option<Image>, ImagesDbError> {
    let record = sqlx::query!(
        r#"
        select id, name, is_default
        from app.images
        where id = $1
        "#,
        image_id,
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| Image {
        id: r.id,
        name: r.name,
        is_default: r.is_default,
    }))
}

pub async fn update_image(
    pool: &PgPool,
    image_id: i64,
    name: &str,
    is_default: bool,
) -> Result<Option<i64>, ImagesDbError> {
    let record = sqlx::query!(
        r#"
        update app.images
        set name = $1, is_default = $2
        where id = $3
        returning id
        "#,
        name,
        is_default,
        image_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_image(pool: &PgPool, image_id: i64) -> Result<Option<i64>, ImagesDbError> {
    let record = sqlx::query!(
        r#"
        delete from app.images
        where id = $1
        returning id
        "#,
        image_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_images(pool: &PgPool) -> Result<Vec<Image>, ImagesDbError> {
    let mut record = sqlx::query!(
        r#"
        select id, name, is_default
        from app.images
        "#,
    )
    .fetch_all(pool)
    .await?;

    Ok(record
        .drain(..)
        .map(|r| Image {
            id: r.id,
            name: r.name,
            is_default: r.is_default,
        })
        .collect())
}

pub async fn read_image_by_replicator_id(
    pool: &PgPool,
    replicator_id: i64,
) -> Result<Option<Image>, ImagesDbError> {
    let record = sqlx::query!(
        r#"
        select i.id, i.name, i.is_default
        from app.images i
        join app.replicators r on i.id = r.image_id
        where r.id = $1
        "#,
        replicator_id,
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| Image {
        id: r.id,
        name: r.name,
        is_default: r.is_default,
    }))
}
