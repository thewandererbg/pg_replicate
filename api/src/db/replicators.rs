use sqlx::{PgPool, Postgres, Transaction};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ReplicatorsDbError {
    #[error("Error while interacting with PostgreSQL for replicators: {0}")]
    Database(#[from] sqlx::Error),
}

pub struct Replicator {
    pub id: i64,
    pub tenant_id: String,
    pub image_id: i64,
}

pub async fn create_replicator(
    pool: &PgPool,
    tenant_id: &str,
    image_id: i64,
) -> Result<i64, ReplicatorsDbError> {
    let mut txn = pool.begin().await?;
    let res = create_replicator_txn(&mut txn, tenant_id, image_id).await;
    txn.commit().await?;
    res
}

pub async fn create_replicator_txn(
    txn: &mut Transaction<'_, Postgres>,
    tenant_id: &str,
    image_id: i64,
) -> Result<i64, ReplicatorsDbError> {
    let record = sqlx::query!(
        r#"
        insert into app.replicators (tenant_id, image_id)
        values ($1, $2)
        returning id
        "#,
        tenant_id,
        image_id
    )
    .fetch_one(&mut **txn)
    .await?;

    Ok(record.id)
}

pub async fn read_replicator_by_pipeline_id(
    pool: &PgPool,
    tenant_id: &str,
    pipeline_id: i64,
) -> Result<Option<Replicator>, ReplicatorsDbError> {
    let record = sqlx::query!(
        r#"
        select r.id, r.tenant_id, r.image_id
        from app.replicators r
        join app.pipelines p on r.id = p.replicator_id
        where r.tenant_id = $1 and p.tenant_id = $1 and p.id = $2
        "#,
        tenant_id,
        pipeline_id,
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| Replicator {
        id: r.id,
        tenant_id: r.tenant_id,
        image_id: r.image_id,
    }))
}

pub async fn read_replicators(
    pool: &PgPool,
    tenant_id: &str,
) -> Result<Vec<Replicator>, ReplicatorsDbError> {
    let mut records = sqlx::query!(
        r#"
        select r.id, r.tenant_id, r.image_id
        from app.replicators r
        join app.pipelines p on r.id = p.replicator_id
        where r.tenant_id = $1 and p.tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(pool)
    .await?;

    Ok(records
        .drain(..)
        .map(|r| Replicator {
            id: r.id,
            tenant_id: r.tenant_id,
            image_id: r.image_id,
        })
        .collect())
}
