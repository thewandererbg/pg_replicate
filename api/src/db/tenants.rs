use sqlx::{PgPool, Postgres, Transaction};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum TenantsDbError {
    #[error("Error while interacting with PostgreSQL for tenants: {0}")]
    Database(#[from] sqlx::Error),
}

pub struct Tenant {
    pub id: String,
    pub name: String,
}

pub async fn create_tenant(
    pool: &PgPool,
    tenant_id: &str,
    tenant_name: &str,
) -> Result<String, TenantsDbError> {
    let mut txn = pool.begin().await?;
    let res = create_tenant_txn(&mut txn, tenant_id, tenant_name).await;
    txn.commit().await?;
    res
}

pub async fn create_tenant_txn(
    txn: &mut Transaction<'_, Postgres>,
    tenant_id: &str,
    tenant_name: &str,
) -> Result<String, TenantsDbError> {
    let record = sqlx::query!(
        r#"
        insert into app.tenants (id, name)
        values ($1, $2)
        returning id
        "#,
        tenant_id,
        tenant_name,
    )
    .fetch_one(&mut **txn)
    .await?;

    Ok(record.id)
}

pub async fn create_or_update_tenant(
    pool: &PgPool,
    tenant_id: &str,
    tenant_name: &str,
) -> Result<String, TenantsDbError> {
    let record = sqlx::query!(
        r#"
        insert into app.tenants (id, name)
        values ($1, $2)
        on conflict (id) do update set name = $2
        returning id
        "#,
        tenant_id,
        tenant_name,
    )
    .fetch_one(pool)
    .await?;

    Ok(record.id)
}

pub async fn read_tenant(pool: &PgPool, tenant_id: &str) -> Result<Option<Tenant>, TenantsDbError> {
    let record = sqlx::query!(
        r#"
        select id, name
        from app.tenants
        where id = $1
        "#,
        tenant_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| Tenant {
        id: r.id,
        name: r.name,
    }))
}

pub async fn update_tenant(
    pool: &PgPool,
    tenant_id: &str,
    tenant_name: &str,
) -> Result<Option<String>, TenantsDbError> {
    let record = sqlx::query!(
        r#"
        update app.tenants
        set name = $1
        where id = $2
        returning id
        "#,
        tenant_name,
        tenant_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_tenant(
    pool: &PgPool,
    tenant_id: &str,
) -> Result<Option<String>, TenantsDbError> {
    let record = sqlx::query!(
        r#"
        delete from app.tenants
        where id = $1
        returning id
        "#,
        tenant_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_tenants(pool: &PgPool) -> Result<Vec<Tenant>, TenantsDbError> {
    let mut record = sqlx::query!(
        r#"
        select id, name
        from app.tenants
        "#,
    )
    .fetch_all(pool)
    .await?;

    Ok(record
        .drain(..)
        .map(|r| Tenant {
            id: r.id,
            name: r.name,
        })
        .collect())
}
