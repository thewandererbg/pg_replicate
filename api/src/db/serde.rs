use serde::de::DeserializeOwned;
use serde::Serialize;
use thiserror::Error;

use crate::encryption::{Decrypt, DecryptionError, Encrypt, EncryptionError, EncryptionKey};

/// Errors that can occur during serialization or encryption for database storage.
#[derive(Debug, Error)]
pub enum DbSerializationError {
    /// Error occurred while serializing data to JSON for the database.
    #[error("Error while serializing data to the db: {0}")]
    Serde(#[from] serde_json::Error),

    /// Error occurred while encrypting data for the database representation.
    #[error("An error occurred while encrypting data for the db representation: {0}")]
    Encryption(#[from] EncryptionError),
}

/// Errors that can occur during deserialization or decryption from database values.
#[derive(Debug, Error)]
pub enum DbDeserializationError {
    /// Error occurred while deserializing data from JSON from the database.
    #[error("Error while deserializing data from the db: {0}")]
    Serde(#[from] serde_json::Error),

    /// Error occurred while decrypting data from the database representation.
    #[error("An error occurred while decrypting data from the db representation: {0}")]
    Decryption(#[from] DecryptionError),
}

/// Serializes a value to a [`serde_json::Value`] for database storage.
///
/// Returns an error if serialization fails.
pub fn serialize<S>(value: S) -> Result<serde_json::Value, DbSerializationError>
where
    S: Serialize,
{
    let serialized_value = serde_json::to_value(value)?;

    Ok(serialized_value)
}

/// Encrypts a value and serializes it to a [`serde_json::Value`] for database storage.
///
/// The value is first encrypted using the provided [`EncryptionKey`], then serialized.
/// Returns an error if encryption or serialization fails.
pub fn encrypt_and_serialize<T, S>(
    value: T,
    encryption_key: &EncryptionKey,
) -> Result<serde_json::Value, DbSerializationError>
where
    T: Encrypt<S>,
    S: Serialize,
{
    let value = value.encrypt(encryption_key)?;
    let serialized_value = serde_json::to_value(value)?;

    Ok(serialized_value)
}

/// Deserializes a [`serde_json::Value`] into a value of type `S`.
///
/// Returns an error if deserialization fails.
pub fn deserialize_from_value<S>(value: serde_json::Value) -> Result<S, DbDeserializationError>
where
    S: DeserializeOwned,
{
    let deserialized_value = serde_json::from_value(value)?;

    Ok(deserialized_value)
}

/// Deserializes and decrypts a [`serde_json::Value`] into a value of type `S`.
///
/// The value is first deserialized into a type implementing [`Decrypt`], then decrypted using
/// the provided [`EncryptionKey`].
///
/// Returns an error if deserialization or decryption fails.
pub fn decrypt_and_deserialize_from_value<T, S>(
    value: serde_json::Value,
    encryption_key: &EncryptionKey,
) -> Result<S, DbDeserializationError>
where
    T: Decrypt<S>,
    T: DeserializeOwned,
{
    let deserialized_value: T = serde_json::from_value(value)?;
    let value = deserialized_value.decrypt(encryption_key)?;

    Ok(value)
}
