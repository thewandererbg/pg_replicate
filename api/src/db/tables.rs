use serde::{Deserialize, Serialize};
use sqlx::{postgres::PgConnectOptions, Connection, Executor, PgConnection, Row};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum TablesDbError {
    #[error("Error while interacting with PostgreSQL for tables: {0}")]
    Database(#[from] sqlx::Error),
}

#[derive(Serialize, Deserialize)]
pub struct Table {
    pub schema: String,
    pub name: String,
}

pub async fn get_tables(options: &PgConnectOptions) -> Result<Vec<Table>, TablesDbError> {
    let mut connection = PgConnection::connect_with(options).await?;
    let query = r#"
        select
           	n.nspname as schema,
           	c.relname as name
        from pg_catalog.pg_class c
           	left join pg_catalog.pg_namespace n on n.oid = c.relnamespace
            left join pg_catalog.pg_am am on am.oid = c.relam
        where
           	c.relkind = 'r'
           	and n.nspname <> 'pg_catalog'
            and n.nspname !~ '^pg_toast'
            and n.nspname <> 'information_schema'
           	and pg_catalog.pg_table_is_visible(c.oid)
        order by schema, name;
        "#;
    let tables = connection
        .fetch_all(query)
        .await?
        .iter()
        .map(|r| Table {
            schema: r.get("schema"),
            name: r.get("name"),
        })
        .collect();
    Ok(tables)
}
