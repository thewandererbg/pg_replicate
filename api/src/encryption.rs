use aws_lc_rs::{
    aead::{Aad, Nonce, RandomizedNonceKey, AES_256_GCM},
    rand::fill,
};
use base64::prelude::BASE64_STANDARD;
use base64::Engine;
use serde::{Deserialize, Serialize};
use std::string;
use thiserror::Error;

/// Errors that can occur during encryption operations.
#[derive(Debug, Error)]
pub enum EncryptionError {
    /// An unspecified error occurred while encrypting data.
    #[error("An unspecified error occurred while encrypting data")]
    Unspecified(#[from] aws_lc_rs::error::Unspecified),
}

/// Errors that can occur during decryption operations.
#[derive(Debug, Error)]
pub enum DecryptionError {
    /// An unspecified error occurred while decrypting data.
    #[error("An unspecified error occurred while decrypting data")]
    Unspecified(#[from] aws_lc_rs::error::Unspecified),

    /// Failed to decode base64 data during decryption.
    #[error("An error occurred while decoding BASE64 data for decryption: {0}")]
    Decode(#[from] base64::DecodeError),

    /// Failed to convert decrypted bytes to UTF-8 string.
    #[error("An error occurred while converting bytes to UTF-8 for decryption: {0}")]
    FromUtf8(#[from] string::FromUtf8Error),

    /// The key ID in the encrypted data did not match the expected key ID.
    #[error("There was a mismatch in the key id while decrypting data (got: {0}, expected: {1})")]
    MismatchedKeyId(u32, u32),
}

/// Trait for types that can be encrypted into another type.
pub trait Encrypt<T> {
    /// Encrypts `self` using the provided [`EncryptionKey`].
    fn encrypt(self, encryption_key: &EncryptionKey) -> Result<T, EncryptionError>;
}

/// Trait for types that can be decrypted into another type.
pub trait Decrypt<T> {
    /// Decrypts `self` using the provided [`EncryptionKey`].
    fn decrypt(self, encryption_key: &EncryptionKey) -> Result<T, DecryptionError>;
}

/// Holds an encryption key and its identifier.
pub struct EncryptionKey {
    /// Unique identifier for the key.
    pub id: u32,
    /// The key material used for encryption and decryption.
    pub key: RandomizedNonceKey,
}

/// Represents an encrypted value with its key ID and nonce.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct EncryptedValue {
    /// Identifier of the key used for encryption.
    pub id: u32,
    /// Base64-encoded nonce used during encryption.
    pub nonce: String,
    /// Base64-encoded encrypted value.
    pub value: String,
}

/// Encrypts a string value using the provided [`EncryptionKey`].
///
/// The result is an [`EncryptedValue`] containing the key ID, base64-encoded nonce,
/// and base64-encoded ciphertext.
pub fn encrypt_text(
    value: String,
    encryption_key: &EncryptionKey,
) -> Result<EncryptedValue, EncryptionError> {
    let (encrypted_password, nonce) = encrypt(value.as_bytes(), &encryption_key.key)?;
    let encoded_encrypted_password = BASE64_STANDARD.encode(encrypted_password);
    let encoded_nonce = BASE64_STANDARD.encode(nonce.as_ref());

    Ok(EncryptedValue {
        id: encryption_key.id,
        nonce: encoded_nonce,
        value: encoded_encrypted_password,
    })
}

/// Decrypts an [`EncryptedValue`] using the provided [`EncryptionKey`].
///
/// Returns the original string if decryption succeeds. Fails if the key ID does not match or if
/// decoding or decryption fails.
pub fn decrypt_text(
    encrypted_value: EncryptedValue,
    encryption_key: &EncryptionKey,
) -> Result<String, DecryptionError> {
    if encrypted_value.id != encryption_key.id {
        return Err(DecryptionError::MismatchedKeyId(
            encrypted_value.id,
            encryption_key.id,
        ));
    }

    let encrypted_value_bytes = BASE64_STANDARD.decode(encrypted_value.value)?;
    let nonce = Nonce::try_assume_unique_for_key(&BASE64_STANDARD.decode(encrypted_value.nonce)?)?;

    let decrypted_value_bytes = decrypt(encrypted_value_bytes, nonce, &encryption_key.key)?;

    let decrypted_value = String::from_utf8(decrypted_value_bytes)?;

    Ok(decrypted_value)
}

/// Encrypts a byte slice using the given [`RandomizedNonceKey`].
///
/// Returns the ciphertext and the nonce used for encryption.
fn encrypt(
    plaintext: &[u8],
    key: &RandomizedNonceKey,
) -> Result<(Vec<u8>, Nonce), aws_lc_rs::error::Unspecified> {
    let mut in_out = plaintext.to_vec();
    let nonce = key.seal_in_place_append_tag(Aad::empty(), &mut in_out)?;

    Ok((in_out, nonce))
}

/// Decrypts a ciphertext using the given [`RandomizedNonceKey`] and [`Nonce`].
///
/// Returns the decrypted plaintext bytes.
fn decrypt(
    mut ciphertext: Vec<u8>,
    nonce: Nonce,
    key: &RandomizedNonceKey,
) -> Result<Vec<u8>, aws_lc_rs::error::Unspecified> {
    let plaintext = key.open_in_place(nonce, Aad::empty(), &mut ciphertext)?;

    Ok(plaintext.to_vec())
}

/// Generates a random [`RandomizedNonceKey`] of length `T` bytes for use with AES-256-GCM.
///
/// The key is filled with cryptographically secure random bytes.
///
/// # Panics
///
/// Panics if `T` does not match the required key length for the cipher.
pub fn generate_random_key<const T: usize>(
) -> Result<RandomizedNonceKey, aws_lc_rs::error::Unspecified> {
    let mut key_bytes = [0u8; T];
    fill(&mut key_bytes)?;

    let key = RandomizedNonceKey::new(&AES_256_GCM, &key_bytes)?;

    Ok(key)
}
