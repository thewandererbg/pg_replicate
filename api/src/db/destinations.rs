use config::shared::DestinationConfig;
use serde::{Deserialize, Serialize};
use sqlx::{PgPool, Postgres, Transaction};
use std::fmt::Debug;
use thiserror::Error;

use crate::db::serde::{
    decrypt_and_deserialize_from_value, encrypt_and_serialize, DbDeserializationError,
    DbSerializationError,
};
use crate::encryption::{
    decrypt_text, encrypt_text, Decrypt, DecryptionError, Encrypt, EncryptedValue, EncryptionError,
    EncryptionKey,
};

impl Encrypt<EncryptedDestinationConfig> for DestinationConfig {
    fn encrypt(
        self,
        encryption_key: &EncryptionKey,
    ) -> Result<EncryptedDestinationConfig, EncryptionError> {
        match self {
            Self::Memory => Ok(EncryptedDestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
            } => {
                let encrypted_service_account_key =
                    encrypt_text(service_account_key, encryption_key)?;

                Ok(EncryptedDestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key: encrypted_service_account_key,
                    max_staleness_mins,
                })
            }
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EncryptedDestinationConfig {
    Memory,
    BigQuery {
        project_id: String,
        dataset_id: String,
        service_account_key: EncryptedValue,
        #[serde(skip_serializing_if = "Option::is_none")]
        max_staleness_mins: Option<u16>,
    },
}

impl Decrypt<DestinationConfig> for EncryptedDestinationConfig {
    fn decrypt(self, encryption_key: &EncryptionKey) -> Result<DestinationConfig, DecryptionError> {
        match self {
            Self::Memory => Ok(DestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key: encrypted_service_account_key,
                max_staleness_mins,
            } => {
                let service_account_key =
                    decrypt_text(encrypted_service_account_key, encryption_key)?;

                Ok(DestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key,
                    max_staleness_mins,
                })
            }
        }
    }
}

#[derive(Debug, Error)]
pub enum DestinationsDbError {
    #[error("Error while interacting with PostgreSQL for destinations: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing destination config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error("Error while deserializing destination config: {0}")]
    DbDeserialization(#[from] DbDeserializationError),
}

pub struct Destination {
    pub id: i64,
    pub tenant_id: String,
    pub name: String,
    pub config: DestinationConfig,
}

pub async fn create_destination(
    pool: &PgPool,
    tenant_id: &str,
    name: &str,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<i64, DestinationsDbError> {
    let config = encrypt_and_serialize(config, encryption_key)?;

    let mut txn = pool.begin().await?;
    let res = create_destination_txn(&mut txn, tenant_id, name, config).await;
    txn.commit().await?;

    res
}

pub async fn create_destination_txn(
    txn: &mut Transaction<'_, Postgres>,
    tenant_id: &str,
    name: &str,
    config: serde_json::Value,
) -> Result<i64, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        insert into app.destinations (tenant_id, name, config)
        values ($1, $2, $3)
        returning id
        "#,
        tenant_id,
        name,
        config
    )
    .fetch_one(&mut **txn)
    .await?;

    Ok(record.id)
}

pub async fn read_destination(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
    encryption_key: &EncryptionKey,
) -> Result<Option<Destination>, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1 and id = $2
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_optional(pool)
    .await?;

    let destination = match record {
        Some(record) => {
            let config = decrypt_and_deserialize_from_value::<
                EncryptedDestinationConfig,
                DestinationConfig,
            >(record.config, encryption_key)?;

            let destination = Destination {
                id: record.id,
                tenant_id: record.tenant_id,
                name: record.name,
                config,
            };

            Some(destination)
        }
        None => None,
    };

    Ok(destination)
}

pub async fn update_destination(
    pool: &PgPool,
    tenant_id: &str,
    name: &str,
    destination_id: i64,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<Option<i64>, DestinationsDbError> {
    let config = encrypt_and_serialize(config, encryption_key)?;

    let mut txn = pool.begin().await?;
    let res = update_destination_txn(&mut txn, tenant_id, name, destination_id, config).await;
    txn.commit().await?;

    res
}

pub async fn update_destination_txn(
    txn: &mut Transaction<'_, Postgres>,
    tenant_id: &str,
    name: &str,
    destination_id: i64,
    config: serde_json::Value,
) -> Result<Option<i64>, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        update app.destinations
        set config = $1, name = $2
        where tenant_id = $3 and id = $4
        returning id
        "#,
        config,
        name,
        tenant_id,
        destination_id
    )
    .fetch_optional(&mut **txn)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_destination(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
) -> Result<Option<i64>, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        delete from app.destinations
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        destination_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_destinations(
    pool: &PgPool,
    tenant_id: &str,
    encryption_key: &EncryptionKey,
) -> Result<Vec<Destination>, DestinationsDbError> {
    let records = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(pool)
    .await?;

    let mut destinations = Vec::with_capacity(records.len());
    for record in records {
        let config = decrypt_and_deserialize_from_value::<
            EncryptedDestinationConfig,
            DestinationConfig,
        >(record.config.clone(), encryption_key)?;

        let destination = Destination {
            id: record.id,
            tenant_id: record.tenant_id,
            name: record.name,
            config,
        };
        destinations.push(destination);
    }

    Ok(destinations)
}

pub async fn destination_exists(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
) -> Result<bool, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        select exists (select id
        from app.destinations
        where tenant_id = $1 and id = $2) as "exists!"
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_one(pool)
    .await?;

    Ok(record.exists)
}

#[cfg(test)]
mod tests {
    use aws_lc_rs::aead::RandomizedNonceKey;
    use config::shared::DestinationConfig;

    use crate::db::destinations::EncryptedDestinationConfig;
    use crate::db::serde::{decrypt_and_deserialize_from_value, encrypt_and_serialize};
    use crate::encryption::EncryptionKey;

    #[test]
    pub fn destination_config_json_deserialization() {
        let config = r#"{
            "big_query": {
                "project_id": "project-id",
                "dataset_id": "dataset-id",
                "service_account_key": "service-account-key",
                "max_staleness_mins": 42
            }
        }"#;

        let deserialized = serde_json::from_str::<DestinationConfig>(config);
        insta::assert_debug_snapshot!(deserialized.unwrap());
    }

    #[test]
    pub fn destination_config_json_serialization() {
        let config = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: "service-account-key".to_string(),
            max_staleness_mins: Some(42),
        };

        insta::assert_json_snapshot!(config);
    }

    #[test]
    pub fn destination_config_json_encryption() {
        let key_bytes = [42u8; 32];
        let key = RandomizedNonceKey::new(&aws_lc_rs::aead::AES_256_GCM, &key_bytes).unwrap();
        let encryption_key = EncryptionKey { id: 1, key };

        let config = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: "supersecretkey".to_string(),
            max_staleness_mins: Some(99),
        };

        let config_in_db = encrypt_and_serialize::<DestinationConfig, EncryptedDestinationConfig>(
            config.clone(),
            &encryption_key,
        )
        .unwrap();
        insta::assert_json_snapshot!(config_in_db, {
            ".big_query.service_account_key" => "[key]"
        });

        let deserialized_config = decrypt_and_deserialize_from_value::<
            EncryptedDestinationConfig,
            DestinationConfig,
        >(config_in_db, &encryption_key)
        .unwrap();
        insta::assert_debug_snapshot!(deserialized_config);
    }
}
