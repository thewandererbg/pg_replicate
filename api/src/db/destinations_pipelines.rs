use config::shared::DestinationConfig;
use sqlx::PgPool;
use thiserror::Error;

use crate::db::destinations::{
    create_destination_txn, update_destination_txn, DestinationsDbError,
};
use crate::db::pipelines::{
    create_pipeline_txn, update_pipeline_txn, PipelineConfig, PipelinesDbError,
};
use crate::db::serde::{
    encrypt_and_serialize, serialize, DbDeserializationError, DbSerializationError,
};
use crate::encryption::EncryptionKey;

#[derive(Debug, Error)]
pub enum DestinationPipelinesDbError {
    #[error("Error while interacting with PostgreSQL for destination and/or pipelines: {0}")]
    Database(#[from] sqlx::Error),

    #[error("The destination with id {0} was not found")]
    DestinationNotFound(i64),

    #[error("The pipeline with id {0} was not found")]
    PipelineNotFound(i64),

    #[error(transparent)]
    PipelinesDb(#[from] PipelinesDbError),

    #[error(transparent)]
    DestinationsDb(#[from] DestinationsDbError),

    #[error("Error while serializing destination or pipeline config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error("Error while deserializing destination or pipeline config: {0}")]
    DbDeserialization(#[from] DbDeserializationError),
}

#[expect(clippy::too_many_arguments)]
pub async fn create_destination_and_pipeline(
    pool: &PgPool,
    tenant_id: &str,
    source_id: i64,
    destination_name: &str,
    destination_config: DestinationConfig,
    image_id: i64,
    pipeline_config: PipelineConfig,
    encryption_key: &EncryptionKey,
) -> Result<(i64, i64), DestinationPipelinesDbError> {
    let destination_config = encrypt_and_serialize(destination_config, encryption_key)?;
    let publication_name = pipeline_config.publication_name.clone();
    let pipeline_config = serialize(pipeline_config)?;

    let mut txn = pool.begin().await?;
    let destination_id =
        create_destination_txn(&mut txn, tenant_id, destination_name, destination_config).await?;
    let pipeline_id = create_pipeline_txn(
        &mut txn,
        tenant_id,
        source_id,
        destination_id,
        image_id,
        &publication_name,
        pipeline_config,
    )
    .await?;
    txn.commit().await?;

    Ok((destination_id, pipeline_id))
}

#[expect(clippy::too_many_arguments)]
pub async fn update_destination_and_pipeline(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
    pipeline_id: i64,
    source_id: i64,
    destination_name: &str,
    destination_config: DestinationConfig,
    pipeline_config: PipelineConfig,
    encryption_key: &EncryptionKey,
) -> Result<(), DestinationPipelinesDbError> {
    let destination_config = encrypt_and_serialize(destination_config, encryption_key)?;
    let publication_name = pipeline_config.publication_name.clone();
    let pipeline_config = serialize(pipeline_config)?;

    let mut txn = pool.begin().await?;
    let destination_id_res = update_destination_txn(
        &mut txn,
        tenant_id,
        destination_name,
        destination_id,
        destination_config,
    )
    .await?;
    if destination_id_res.is_none() {
        txn.rollback().await?;
        return Err(DestinationPipelinesDbError::DestinationNotFound(
            destination_id,
        ));
    };
    let pipeline_id_res = update_pipeline_txn(
        &mut txn,
        tenant_id,
        pipeline_id,
        source_id,
        destination_id,
        &publication_name,
        pipeline_config,
    )
    .await?;

    if pipeline_id_res.is_none() {
        txn.rollback().await?;
        return Err(DestinationPipelinesDbError::PipelineNotFound(pipeline_id));
    };

    txn.commit().await?;

    Ok(())
}
