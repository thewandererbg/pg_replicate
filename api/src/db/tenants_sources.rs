use sqlx::PgPool;
use thiserror::Error;

use crate::db::serde::{encrypt_and_serialize, DbSerializationError};
use crate::db::sources::{create_source_txn, SourceConfig, SourcesDbError};
use crate::db::tenants::{create_tenant_txn, TenantsDbError};
use crate::encryption::EncryptionKey;

#[derive(Debug, Error)]
pub enum TenantSourceDbError {
    #[error("Error while interacting with PostgreSQL for tenants and/or sources: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing tenant or source config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error(transparent)]
    Sources(#[from] SourcesDbError),

    #[error(transparent)]
    Tenants(#[from] TenantsDbError),
}

pub async fn create_tenant_and_source(
    pool: &PgPool,
    tenant_id: &str,
    tenant_name: &str,
    source_name: &str,
    source_config: SourceConfig,
    encryption_key: &EncryptionKey,
) -> Result<(String, i64), TenantSourceDbError> {
    let source_config = encrypt_and_serialize(source_config, encryption_key)?;

    let mut txn = pool.begin().await?;
    let tenant_id = create_tenant_txn(&mut txn, tenant_id, tenant_name).await?;
    let source_id = create_source_txn(&mut txn, &tenant_id, source_name, source_config).await?;
    txn.commit().await?;

    Ok((tenant_id, source_id))
}
