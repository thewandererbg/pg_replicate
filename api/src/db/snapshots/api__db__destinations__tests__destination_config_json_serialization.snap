---
source: api/src/db/destinations.rs
expression: config
---
{
  "big_query": {
    "project_id": "project-id",
    "dataset_id": "dataset-id",
    "service_account_key": "service-account-key",
    "max_staleness_mins": 42
  }
}
